File: app/backend/app/api/session.py
from flask import Blueprint, request, make_response
from flask_login import login_user, logout_user
from werkzeug.datastructures import MultiDict
from flask_wtf.csrf import generate_csrf
from ..forms.login import LoginForm
from ..models import User

session = Blueprint('session', __name__)


@session.route('/login', methods=['POST'])
def login():
    data = MultiDict(mapping=request.json)
    form = LoginForm(data)
    if form.validate():
        user = User.query.filter(User.username == 'kloie').first()
        if user and user.checkPassword(data['password']):
            login_user(user)
            userLogged = user.to_dict()
            return {'user': userLogged}
        else:
            res = make_response({'errors': ['User does not exist']}, 401)
            return res
    else:
        res = make_response(
            {'errors': [form.errors[error][0] for error in form.errors]}, 401)
        return res


@session.route('/logout', methods=['DELETE'])
def logout():
    logout_user()
    return {'message': 'Logged out'}


@session.route('/csrf')
def csrf():
    res = make_response('set token')
    res.set_cookie('XSRF-TOKEN', generate_csrf())
    return res

File: app/frontend/src/store/session.js
import Cookies from 'js-cookie';


const SET_SESSION = 'session/SET_SESSION';
const LOGOUT_USER = 'session/LOGOUT_USER';

export const setSession = user => {
    return {
        type: SET_SESSION,
        user
    }
};

export const logoutUser = () => {
    return {
        type: LOGOUT_USER
    }
};

export const login = password => {
    const csrfToken = Cookies.get('XSRF-TOKEN');

    return async dispatch => {
        const res = await fetch('/api/session/login', {
            method: 'post',
            headers: {
                'Content-Type': 'application/json',
                'X-CSRFTOKEN': csrfToken
            },
            body: JSON.stringify({ password, 'csrf_token': csrfToken })
        });
        res.data = await res.json();

        if (res.ok) {
            dispatch(setSession(res.data.user));
        }
        return res;
    }
};

export const logout = () => {
    const csrfToken = Cookies.get('XSRF-TOKEN');
    return async dispatch => {
        const res = await fetch('/api/session/logout', {
            method: 'delete',
            headers: {
                'X-CSRFTOKEN': csrfToken
            }
        });
        res.data = await res.json();

        if (res.ok) {
            dispatch(logoutUser());
        }
        return res;
    }
};

const initState = {
    user: null
};

export default function sessionReducer(state = initState, action) {
    const newState = Object.assign({}, state);
    switch (action.type) {
        case SET_SESSION:
            newState.user = action.user;
            return newState;
        case LOGOUT_USER:
            return {};
        default:
            return state;
    }
};